lib/DBIx/TransactionManager/Distributed.pm

```
@@ -0,0 +1,268 @@
+Expects a category (string value) and L<DBI::db> instance.
+
+Returns the database handle.
+
+Example:
+
+    sub _dbh {
+        my $dbh = DBI->connect_cached('dbi:Pg', '', '', { RaiseError => 1});
+        return register_cached_dbh('category' => $dbh);
+    }
+
+=cut
+
+sub register_cached_dbh {
+    my ($category, $dbh) = @_;
+    register_dbh($category => $dbh) unless BOM::Database::dbh_is_registered($category => $dbh);
+    $dbh->begin_work if $IN_TRANSACTION && $dbh->{AutoCommit};
+    $dbh;
+}
+
+=head2 txn
+
+Runs the given coderef in a transaction.
+
+Expects a coderef and one or more database handle categories.
+
+Will call L<DBI/begin_work> for every known database handle in the given category,
+run the code, then call L<DBI/commit> on success, or L<DBI/rollback> on failure.
+
+Will raise an exception on failure, or return an empty list on success.
+
+Example:
+
+    txn { dbh()->do('NOTIFY something') } 'category';
+
+WARNING: This only applies transactions to known database handles. Anything else -
+Redis, cache layers, files on disk - is out of scope. Transactions are a simple
+L<DBI/begin_work> / L<DBI/commit> pair, there's no 2-phase commit or other
+distributed transaction co-ordination happening here.
+
+=cut
+
+sub txn(&;@) {
+    my ($code, @categories) = @_;
+    die "Need a database category" unless @categories;
+    _check_fork();
+    my $wantarray = wantarray;
+    for my $category (@categories) {
+        if (my $count = () = extract_by { !defined($_) } @{$DBH{$category}}) {
+            warn "Had $count database handles that were not released via release_dbh, probable candidates follow:\n";
+            my %addr = map { ; refaddr($_) => 1 } @{$DBH{$category}};
+            warn "unreleased dbh in $_\n" for sort delete @{$DBH_SOURCE{$category}}{grep !exists $addr{$_}, keys %{$DBH_SOURCE{$category}}};
+        }
+    }
+
+    my @rslt;
+    eval {
+        for my $category (@categories) {
+            $_->begin_work for @{$DBH{$category}};
```

@binarytom  Should we also check `$dbh->{AutoCommit}` here ? Otherwise it will emit an error if it is already set.

---

Yes, good idea.

---

@binarytom  We used `RaiseError => 1` attribute, so should we test this attribute when register the dbh ? Or we set its value locally here ? `local $_->{RaiseError} = 1`

---

@chylli We could raise an exception if it's not set - since the flag can be changed at any time, we'd need to check before the `->begin_work` call and again before `->commit`, not sure if it's worth the effort. Could just document that the module expects `RaiseError` to be set - up to you.